verifier/verifier.go

```diff
@@ -0,0 +1,153 @@
+package verifier
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type ProviderVerifier interface {
+	// Returns the OpenID provider issuer as seen in ID token e.g. "https://accounts.google.com"
+	Issuer() string
+	ProviderPublicKey(ctx context.Context, token []byte) (JSONWebKey, error)
+	VerifyProvider(ctx context.Context, pkt *pktoken.PKToken) error
+}
+
+type CosignerVerifier interface {
+	Issuer() string
+	Strict() bool // Whether or not a given cosigner MUST be present for successful verification
+	VerifyCosigner(ctx context.Context, pkt *pktoken.PKToken) error
+}
+
+type VerifierOpts func(*Verifier) error
+
+func WithCosignerVerifiers(verifiers ...*DefaultCosignerVerifier) VerifierOpts {
+	return func(v *Verifier) error {
+		for _, verifier := range verifiers {
+			fmt.Println(verifier.issuer)
+			v.cosigners[verifier.issuer] = verifier
+		}
+		return nil
+	}
+}
+
+func AddProviderVerifiers(verifiers ...ProviderVerifier) VerifierOpts {
+	return func(v *Verifier) error {
+		for _, verifier := range verifiers {
+			if _, ok := v.providers[verifier.Issuer()]; ok {
+				return fmt.Errorf("provider verifier found with duplicate issuer: %s", verifier.Issuer())
+			}
+			v.providers[verifier.Issuer()] = verifier
+		}
+		return nil
+	}
+}
+
+type Verifier struct {
+	providers map[string]ProviderVerifier
+	cosigners map[string]CosignerVerifier
+}
+
+func New(verifier ProviderVerifier, options ...VerifierOpts) (*Verifier, error) {
+	v := &Verifier{
+		providers: map[string]ProviderVerifier{
+			verifier.Issuer(): verifier,
+		},
+		cosigners: map[string]CosignerVerifier{},
+	}
+
+	for _, option := range options {
+		if err := option(v); err != nil {
+			return nil, err
+		}
+	}
+
+	return v, nil
+}
+
+// Verifies whether a PK token is valid and matches all expected claims.
+//
+// extraChecks: Allows for optional specification of additional checks
+func (v *Verifier) VerifyPKToken(
+	ctx context.Context,
+	pkt *pktoken.PKToken,
+	extraChecks ...Check,
+) error {
+	issuer, err := pkt.Issuer()
+	if err != nil {
+		return err
+	}
+
+	providerVerfier, ok := v.providers[issuer]
```

Comment:
Can you add a comment similar to my comment here which specifies the current limitation of providers with different audiences and the same issuer. I know your check above would cause an error but we do eventually want to support the same issuer with different audiences

https://github.com/openpubkey/openpubkey/blob/main/client/oidc.go#L125C1-L133C1